custom/src/main/java/com/splunk/opentelemetry/middleware/MiddlewareAttributeSpanProcessor.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.middleware;
+
+import com.splunk.opentelemetry.javaagent.bootstrap.MiddlewareHolder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.trace.ReadWriteSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+
+public class MiddlewareAttributeSpanProcessor implements SpanProcessor {
+
+  @Override
+  public void onStart(Context parentContext, ReadWriteSpan span) {
+    String middlewareName = MiddlewareHolder.middlewareName.get();
+    String middlewareVersion = MiddlewareHolder.middlewareVersion.get();
+    // Getting span kind is not the most straightforward or cheap operation apparently.
+    // Let's do quick cheap null checks first and exit quickly.
+    if ((middlewareName == null && middlewareVersion == null)
+        || span.toSpanData().getKind() != Span.Kind.SERVER) {
```

Comment:
This is bad :( You really should not convert span to SpanData during `onStart`, this conversion is potentially quite expensive.

@bogdandrutu you objected on numerous occasions against premature span conversion/actualisation. Can you point us to some explanation why this is bad?

@jkwatson any reasons why we cannot add `getSpanKind` to `ReadableSpan`?

Reply:
I know this is bad, and I already asked the same question in gitter.

Reply:
https://gitter.im/open-telemetry/community?at=5fc0b99be81def77d84c300d

Reply:
Here's the PR to SDK for making getKind public: https://github.com/open-telemetry/opentelemetry-java/pull/2162

@iNikem Should I add a TODO here to update once the PR is merged and released?

Reply:
Yes, please